testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/userdata/ContactListActivity.kt

```diff
@@ -0,0 +1,149 @@
+import com.rakuten.tech.mobile.testapp.ui.base.BaseActivity
+import com.rakuten.tech.mobile.testapp.ui.settings.AppSettings
+import com.rakuten.tech.mobile.testapp.ui.settings.SettingsMenuActivity
+import kotlinx.android.synthetic.main.contacts_activity.*
+import java.util.UUID
+
+class ContactListActivity : BaseActivity() {
+
+    private lateinit var settings: AppSettings
+    private val adapter = ContactListAdapter()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        settings = AppSettings.instance
+        setContentView(R.layout.contacts_activity)
+        initializeActionBar()
+        renderContactList()
+        fabAddContact.setOnClickListener { onAddAction() }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        menuInflater.inflate(R.menu.settings_menu, menu)
+        return true
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            android.R.id.home -> {
+                finish()
+                return true
+            }
+            R.id.settings_menu_save -> {
+                onSaveAction()
+                return true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    private fun onSaveAction() {
+        settings.contactNames = adapter.provideContactEntries()
+        finish()
+    }
+
+    private fun onAddAction() {
+        val contactView = layoutInflater.inflate(R.layout.dialog_add_contact, null)
+        val editNewContact: AppCompatEditText = contactView.findViewById(R.id.editNewContact)
+        ContactInputDialog.Builder().build(this).apply {
+            setView(contactView)
+            setPositiveListener(DialogInterface.OnClickListener { dialog, _ ->
+                if (editNewContact.text.toString().isNotEmpty()) {
+                    val name = editNewContact.text.toString()
+                    // add contact in the last position
+                    adapter.addContact(adapter.itemCount, clearWhiteSpaces(name))
+                } else {
+                    Toast.makeText(
+                        this@ContactListActivity,
+                        getString(R.string.userdata_error_invalid_contact),
+                        Toast.LENGTH_LONG
+                    ).apply {
+                        setGravity(Gravity.BOTTOM, 0, 100)
+                    }.show()
+                }
+                dialog.cancel()
+            })
+        }.show()
+    }
+
+    private fun initializeActionBar() {
+        val toolBar = findViewById<Toolbar>(R.id.toolbar)
+        setSupportActionBar(toolBar)
+        showBackIcon()
+    }
+
+    private fun renderContactList() {
+        var contactNames = ArrayList<String>()
+        if (settings.contactNames.isEmpty())
+            contactNames = createRandomUUIDList()
+        else
+            contactNames.addAll(settings.contactNames)
```

Comment:
When the user removes all contacts, it creates the contact list again. I don't have strong opinion on this but in user perspective, it is bug.